nasa/ow_autonomy
# Fault Handling Pattern 3

This plan that executes a very short, simple set of behaviors (i.e.. just pan 180 degrees, unstow and stow the arm) and pauses when any faults are injected, and resumes when they are removed. This plan is basically a simple version of ReferenceMission2.plp.

Important notes: 
- There is a bug with antenna faults right now, where the antenna does not stop rotating even when an antenna fault is injected. This is not related to how this plan is designed and will be fixed in melodic-devel.
- When a fault is injected during the execution of an action, only the relevant kind of fault will halt that action immediately (i.e. arm fault for unstow or stow, since antenna fault reactions need a bug fix). Otherwise the current action will finish, and the next action will then not start until there are no faults in the system. There is a bug related to this, see ** below -- Resolved through avoidance by changing SystemFault consideration to either AntennaFault or ArmFault...

To test:
- To see the issue with antenna fault reactivity, wait for initialization to complete and the 180 pan to start and inject an antenna fault. What should happen is the antenna should halt and then the plan should execute the 10-second pausing. Instead, panning will complete and then the pausing should occur (until the fault is removed).
- Try injecting a fault while unstow is occurring. 
    - If you inject an arm fault, you should see the unstow halt and the pausing commence until you remove the fault. When you remove it, stowing should begin, regardless of whether unstow finished.
    - ** If you inject an arm fault AND an antenna fault, you should see the same behavior as the previous. However, try removing the arm fault. Since there should still be a system fault (antenna is part of the system), stowing should NOT start. Sometimes though, it does. I think this is a bug (Mike is working on it). **
- Try injecting a fault during stowing. 
    - If you inject an arm fault, stowing should stop, and once you remove the fault the plan will be complete. 
    - If you inject an antenna fault the stowing should still complete since this plan does not deal with injection *during* node execution directly (unless the fault is inherently connected, i.e. arm fault with arm action).
